crates/providers/src/heart.rs

```diff
@@ -0,0 +1,276 @@
+use futures::{stream::StreamExt, FutureExt, Stream};
+use std::{
+    collections::{BTreeMap, HashMap},
+    fmt,
+    future::Future,
+    time::{Duration, Instant},
+};
+use tokio::{
+    select,
+    sync::{mpsc, oneshot, watch},
+};
+
+/// A configuration object for watching for transaction confirmation.
+pub struct WatchConfig {
+    /// The transaction hash to watch for.
+    tx_hash: B256,
+
+    /// Require a number of confirmations.
+    confirmations: u64,
+
+    /// Optional timeout for the transaction.
+    timeout: Option<Duration>,
+}
+
+impl WatchConfig {
+    /// Create a new watch for a transaction.
+    pub fn new(tx_hash: B256) -> Self {
+        Self { tx_hash, confirmations: 0, timeout: None }
+    }
+
+    /// Set the number of confirmations to wait for.
+    pub fn set_confirmations(&mut self, confirmations: u64) {
+        self.confirmations = confirmations;
+    }
+
+    /// Set the number of confirmations to wait for.
+    pub fn with_confirmations(mut self, confirmations: u64) -> Self {
+        self.confirmations = confirmations;
+        self
+    }
+
+    /// Set the timeout for the transaction.
+    pub fn set_timeout(&mut self, timeout: Duration) {
+        self.timeout = Some(timeout);
+    }
+
+    /// Set the timeout for the transaction.
+    pub fn with_timeout(mut self, timeout: Duration) -> Self {
+        self.timeout = Some(timeout);
+        self
+    }
+}
+
+struct TxWatcher {
+    config: WatchConfig,
+    tx: oneshot::Sender<()>,
+}
+
+impl TxWatcher {
+    /// Notify the waiter.
+    fn notify(self) {
+        let _ = self.tx.send(());
+    }
+}
+
+/// A pending transaction that can be awaited.
+pub struct PendingTransaction {
+    /// The transaction hash.
+    pub(crate) tx_hash: B256,
+    /// The receiver for the notification.
+    // TODO: send a receipt?
+    pub(crate) rx: oneshot::Receiver<()>,
+}
+
+impl fmt::Debug for PendingTransaction {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("PendingTransaction").field("tx_hash", &self.tx_hash).finish()
+    }
+}
+
+impl PendingTransaction {
+    /// Returns this transaction's hash.
+    pub const fn tx_hash(&self) -> &B256 {
+        &self.tx_hash
+    }
+}
+
+impl Future for PendingTransaction {
+    type Output = TransportResult<()>;
+
+    fn poll(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        self.rx.poll_unpin(cx).map(|res| res.map_err(|_| TransportErrorKind::backend_gone()))
+    }
+}
+
+/// A handle to the heartbeat task.
+#[derive(Debug, Clone)]
+pub(crate) struct HeartbeatHandle {
+    tx: mpsc::Sender<TxWatcher>,
+    latest: watch::Receiver<Option<Block>>,
+}
+
+impl HeartbeatHandle {
+    /// Watch for a transaction to be confirmed with the given config.
+    pub(crate) async fn watch_tx(
+        &self,
+        config: WatchConfig,
```

Comment:
```suggestion
        config: impl Into<WatchConfig>,
```

Reply:
It's not a public function, doesn't really matter